### ISSUE TITLE
We need a component-level diagram to illustrate the Prior Authorization Questionnaire Use Case

### ISSUE THREAD

**Issue by dmuylwyk** [cqframework/cqf-us#25]
We need a component-level diagram to illustrate the Prior Authorization Questionnaire Use Case. This diagram should include:

- components that are in play;
- systems they are part of; and
- what interactions they support with pointers to detailed interactions.

This diagram should appear under section 1.2.1 Prior Authorization Questionnaire Use Case, and it should be accompanied by narrative as necessary.

**Comment by dmuylwyk** [comment]
Hi, @brynrhodes. Are you able to weigh in on the questions below?

This diagram _could_ be as simple as the draft below but I'm not sure it's of any value. Note the draft is informed by expected systems defined [here](https://hl7.org/fhir/us/davinci-dtr/#expected-systems):

![20240824-Draft-Prior-Authorization-Questionnaire-Use-Case](https://github.com/user-attachments/assets/2a3a6982-bf22-4694-8b1d-016554af9a78)

When considering the [sequence diagram available here](https://hl7.org/fhir/us/davinci-dtr/specification.html#formal-specification) (also reproduced below), the only thing that seems to be missing is more detail w.r.t. how the DTR App executes CQL and retrieves data from the EHR. SDC's `$populate` appears to be the answer but it's not explicitly named, and includes some limitations.

Additoinally, SDC doesn't appear to mention the [CQF Library extension](https://hl7.org/fhir/extensions/StructureDefinition-cqf-library.html) at all, which CQF-US requires.

## Bottom Line

I'm not sure how the proposed diagram and narrative should appear. DTR describes form population using Library-based CQL expressions via the [CQF Library extension](https://hl7.org/fhir/extensions/StructureDefinition-cqf-library.html); however, SDC does not. I need to plug this gap in my knowledge.

## Questions/Concerns

### Question 1

**Is the DTR App expected to invoke `$populate` on the Provider EHR?**

If the answer to No.1 is yes, I'm concerned because I cannot find any mention of the CQF Library extension in SDC. The question becomes: Does or will `$populate` support Library-based CQL expressions via the CQF Library extension?

Additionally, the expression language of `text/cql-identifier` was introduced in FHIR R4B, which may be a consideration for Prior Authorization Questionnaire Use Case implementations using FHIR R4.

### Question 2

**Alternatively, is the DTR App expected to implement behaviour similar to `$populate`; being solely responsible for interpreting the Questionnaire, executing the CQL, and retrieving data from the Provider EHR?**

If the answer to No.2 is yes, I'm concerned about having to do CQL execution in the app. This could theoretically be delegated to an intermediary component that exists between the DTR App and the Provider EHR; unfortunately, `$populate` as defined does not support such delegation (i.e. there's no "get your data from _this_ FHIR endpoint" input parameter). This would require enhancing `$populate`, overloading `$populate`, or defining a new operation.

## Knowledge Dump

This initially looked like a relatively simply task. There are lots of related resources at the following links:

- https://hl7.org/fhir/us/davinci-dtr/#expected-systems
- https://hl7.org/fhir/us/davinci-dtr/specification.html#overview
- https://hl7.org/fhir/us/davinci-dtr/epa.html
- https://hl7.org/fhir/us/davinci-dtr/OperationDefinition-questionnaire-package.html
- https://hl7.org/fhir/uv/sdc/populate.html
- https://hl7.org/fhir/uv/sdc/OperationDefinition-Questionnaire-populate.html
- https://hl7.org/fhir/us/davinci-crd/StructureDefinition-ext-coverage-information.html

I've reproduced the sequence diagram [available here](https://hl7.org/fhir/us/davinci-dtr/specification.html#overview), and I've added numbers for reference purposes:

```mermaid
sequenceDiagram
    participant pro as Provider EHR / Authorization Server
    participant dtr as DTR App
    participant pay as Payer
    pro-->dtr: 1. Configure EHR to App connectivity
    pro-->pay: 2. Configure EHR / App to Payer Connectivity
    note over pro: 3. SMART / Native App Launch
    pro->>dtr: 
    note over dtr: 4. DTR App leverages<br/>pre-defined Payer information
    dtr->>pay: 5. DTR's $questionnaire-package operation
    pay-->>dtr: 6. Returns "Questionnaire Package"
    note over dtr: 7. DTR App uses "Questionnaire Package"<br/>to pre-populate QuestionnaireResponse
    dtr->>pro: 8. Execute CQL / Retrieve data from EHR
    note over dtr: 9. User verifies data and<br/>QuestionnaireResponse
    dtr->>pro: 10. Store completed QuestionnaireResponse
```

In steps 4-6, we determine the needed Questionnaire in one of at least two ways:

- a [Coverage Information extension](https://hl7.org/fhir/us/davinci-crd/StructureDefinition-ext-coverage-information.html) is included on the order resource(s) passed via the `order` input parameter of DTR's `$questionnaire-package` operation; and/or
- a context ID is passed via the `context` input parameter of the same operation (e.g. where a PAS Response Bundle includes a request for additional information, and the DTR process needs to occur again).

Unfortunately, I'm a little confused by steps 7-8. The DTR App should use data from both the Provider EHR and the Questionnaire Package returned by the Payer to populate a QuestionnaireResponse, which it stores with the Provider EHR. It seems this should happen using one of SDC's population operations (`$populate`).

DTR describes use of the CQF Library extension [here](https://hl7.org/fhir/us/davinci-dtr/specification.html#:~:text=DTR%20uses%20the,populate%20the%20QuestionnaireResponse.):

> DTR uses the [SDC Expression-based Population mechanism](http://hl7.org/fhir/uv/sdc/STU3/populate.html#expression-based-population) as the mechanism for populating, and specifically the use of CQL as the expression language. CQL SHALL be used when pre-populating elements using data gathered from the EHR. This will mean using the launchContext, [variable](https://hl7.org/fhir/extensions/StructureDefinition-variable.html) and the [initialExpression](http://hl7.org/fhir/uv/sdc/STU3/StructureDefinition-sdc-questionnaire-initialExpression.html), [contextExpression](http://hl7.org/fhir/uv/sdc/STU3/StructureDefinition-sdc-questionnaire-contextExpression.html) or [calculatedExpression](http://hl7.org/fhir/uv/sdc/STU3/StructureDefinition-sdc-questionnaire-calculatedExpression.html) extensions together with logic in [Libraries](http://hl7.org/fhir/R4/library.html) referenced by the [cqf-library](https://hl7.org/fhir/extensions/StructureDefinition-cqf-library.html) extension to help populate the QuestionnaireResponse.

Expression-based population means using one of the population operations defined in SDC. For example, `$populate`:

- https://hl7.org/fhir/uv/sdc/OperationDefinition-Questionnaire-populate.html

**Comment by brynrhodes** [comment]
Thank you for the detailed comments and questions, the short answer is that the DTR specification is not prescriptive about the use of the $populate operation, since the DTR application is effectively responsible for the population. It could choose to use a $populate operation, but it may not, choosing to run the CQL itself directly. I've provided a simple diagram in this branch and identified the main points of the data flow from the perspective of CQL evaluation. I've also identified the contexts in which CQL expressions may be run throughout the DTR process, and documented how that CQL will be parameterized from the various DTR contexts. We will review on the community call tomorrow.
